pkg/ha/ha_service.go

```diff
@@ -179,71 +154,43 @@ func (h *Service) tryChangeLeader(cluster, currentLeader string) {
 	defer clusterLock.Release(1)
 
 	for {
-		ss, err := h.loadState(cluster)
-		if err != nil {
-			log.Error("error", err)
+		leaseView := currentLease.Clone()
+		if ok := s.shouldTryToChangeLeader(leaseView); ok {
 			return
 		}
-		stateView := ss.Clone()
-		if ok := h.checkLeaseUntilAndLastWrite(stateView); ok {
-			return
-		}
-		lockState, err := h.lockClient.tryChangeLeader(
-			context.Background(), cluster, stateView.MaxTimeInstance, stateView.MaxTimeSeen,
+		leaseState, err := s.leaseClient.TryChangeLeader(
+			context.Background(), cluster, leaseView.MaxTimeInstance, leaseView.MaxTimeSeen,
 		)
 		if err != nil {
 			log.Error("msg", "Couldn't change leader", "err", err)
 			return
 		}
 
-		ss.UpdateStateFromDB(lockState)
-		if lockState.Leader != currentLeader {
+		err = currentLease.SetUpdateFromDB(leaseState)
```

Comment:
so the lease UpdateFromDB path has the service pass the client to the lease, which does the operation on the client and internally sets the state. For changing leader we have the service use the client to do the operation and then change the lease state with setUpdateFromDB. I think we should choose one approach or the other but not both. Either the lease executes the operation on the client or the service does. My hunch is the service uses the client and updates the lease but I am not sure, will leave up to you.